### PR TITLE
Display special peers at the top of the list

### PR DESCRIPTION
**Pull request by quantumagi**
See https://stratisplatformuk.visualstudio.com/StratisBitcoinFullNode/_workitems/edit/3416.

### CODE REVIEW

File: src/Stratis.Bitcoin/Connection/ConnectionManager.cs [hunk 1]
```diff
@@ -257,6 +259,41 @@ private void AddComponentStats(StringBuilder builder)
                     + " agent:" + agent);
             }
 
+            var oneTryPeers = this.ConnectedPeers.Where(p => p.Behavior<ConnectionManagerBehavior>().OneTry);
```
Comment (dangershony) [inline src/Stratis.Bitcoin/Connection/ConnectionManager.cs]: Is it a good idea to iterate ConnectedPeers 3 times?

**Comment by dangershony** [comment]
@quantumagi please provide a prt screen

File: src/Stratis.Bitcoin/Connection/ConnectionManager.cs (hunk 1 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin/Connection/ConnectionManager.cs]: In this case I would say its fairly innocuous. Typically there are not that many items so setting up two additional string builders may be overkill. If you want I can easily change it.

**Comment by quantumagi** [comment]
image

File: src/Stratis.Bitcoin/Connection/ConnectionManager.cs (hunk 1 above)
Comment (codingupastorm) [reply src/Stratis.Bitcoin/Connection/ConnectionManager.cs]: `.Behavior<ConnectionManagerBehavior>()` also does a FirstOrDefault() to get the correct behaviour. Considering we could have ~100 peers which could have several Behaviours, and will be doing this every 5 seconds, maybe there is a way we can store them in some kind of mapping so we only do this work once?

Comment (quantumagi) [reply src/Stratis.Bitcoin/Connection/ConnectionManager.cs]: I have made some changes to only iterate ConnectedPeers once. Created a dictionary for AddNode checking. Further optimization is certainly possible if we regard this important enough.

File: src/Stratis.Bitcoin/Connection/ConnectionManager.cs [hunk 2]
```diff
@@ -257,11 +258,69 @@ private void AddComponentStats(StringBuilder builder)
...
+
             int inbound = this.ConnectedPeers.Count(x => x.Inbound);
 
             builder.AppendLine();
             builder.AppendLine($"======Connection====== agent {this.Parameters.UserAgent} [in:{inbound} out:{this.ConnectedPeers.Count() - inbound}] [recv: {totalRead.BytesToMegaBytes()} MB sent: {totalWritten.BytesToMegaBytes()} MB]");
-            builder.AppendLine(peerBuilder.ToString());
+
+            if (oneTryBuilder.Length > 0)
+            {
+                builder.AppendLine(">>> Connect:");
```
Comment (dangershony) [inline src/Stratis.Bitcoin/Connection/ConnectionManager.cs]: This is not correct onetry and connect are different things

Comment (dangershony) [reply src/Stratis.Bitcoin/Connection/ConnectionManager.cs]: its called be called `onetry` ( we probably not need to specify `connect` when that is specified all other options are off (apart for whitelist).

What about cases where a peer is both `addnode` and `whitlisted`?

File: src/Stratis.Bitcoin/Connection/ConnectionManager.cs [hunk 3]
```diff
@@ -257,11 +258,69 @@ private void AddComponentStats(StringBuilder builder)
...
+            }
+
+            if (addNodeBuilder.Length > 0)
+            {
+                builder.AppendLine(">>> AddNode:");
+                builder.Append(addNodeBuilder.ToString());
+                builder.AppendLine("<<<");
+            }
+
+            if (otherBuilder.Length > 0)
```
Comment (dangershony) [inline src/Stratis.Bitcoin/Connection/ConnectionManager.cs]: I suggest we delet the `<<<` and just add here `discovered`

File: src/Stratis.Bitcoin/Connection/ConnectionManager.cs [hunk 4]
```diff
@@ -257,11 +258,69 @@ private void AddComponentStats(StringBuilder builder)
...
+                if (connectionManagerBehavior.OneTry)
+                {
+                    AddPeerInfo(oneTryBuilder, peer);
+                    added = true;
+                }
+
+                if (connectionManagerBehavior.Whitelisted)
+                {
+                    AddPeerInfo(whiteListedBuilder, peer);
+                    added = true;
```
Comment (dangershony) [inline src/Stratis.Bitcoin/Connection/ConnectionManager.cs]: A whitelisted node can also be an addnode so there will be a duplicate (if that is an issue or not I am not sure)

Comment (quantumagi) [reply src/Stratis.Bitcoin/Connection/ConnectionManager.cs]: I would go for simplicity here and perhaps accept that a node may be listed twice.

Comment (dangershony) [reply src/Stratis.Bitcoin/Connection/ConnectionManager.cs]: I can live with that

File: src/Stratis.Bitcoin/Connection/ConnectionManager.cs (hunk 2 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin/Connection/ConnectionManager.cs]: Fixed.

**Pull request merged**